lib/CriticalRSM.cc

```diff
@@ -0,0 +1,290 @@
+#include "Utils.h"
+#include "Vertex.h"
+
+CriticalRSM::CriticalRSM(std::shared_ptr<BipartiteGraph> G, bool A_proposing)
+    : MatchingAlgorithm(G, A_proposing)
+{}
+
+
+Matching CriticalRSM::compute_matching() {
+  FreeListType free_list;
+  std::map<VertexPtr, VertexBookkeeping> bookkeep_data;
+  std::shared_ptr<BipartiteGraph> G = get_graph();
+  auto M = Matching(is_A_proposing());
+  
+  // choose the partitions from which the vertices will propose
+  const auto& proposing_partition = is_A_proposing() ? G->get_A_partition()
+                                                     : G->get_B_partition();
+
+  const auto& non_proposing_partition = is_A_proposing() ? G->get_B_partition()
+                                                         : G->get_A_partition();
+
+  // s is sum of lower quotas of proposing vertices.
+  auto s = 0;
+  for (const auto& [_, v] : proposing_partition) {
+    s += v->get_lower_quota();
+  }
+
+  // t is sum of lower quotas of non-proposing vertices.
+  auto t = 0;
+  for (const auto& [_, v] : non_proposing_partition) {
+    t += v->get_lower_quota();
+  }
+
+  // set the level of every vertex in the proposing partition to 0
+  // mark all proposing vertices free (by pushing into the free_list)
+  // and vertices from the opposite partition implicitly free
+  for (const auto& [_, v] : proposing_partition) {
+    //free_list.push(v);
+    int pref_list_size = v->get_preference_list().size();
+    int pref_list_lq_size = v->get_preference_list_lq().size();
+    auto vertex_data{VertexBookkeeping(0, pref_list_size,
+                                       0, pref_list_lq_size)};
+    bookkeep_data[v] = vertex_data;
+    add_to_free_list(free_list, vertex_data, v);
+  }
+
+  // Make prefSC and prefS for each vertex
+  std::unordered_map<VertexPtr, PreferenceList> prefS;
+  std::unordered_map<VertexPtr, PreferenceList> prefSC;
+  
+  for (auto &it: proposing_partition) {
+    auto a = it.second;
+    auto pref_list_a = a->get_preference_list();
+    prefS[a] = pref_list_a.get_prefS();
+    prefSC[a] = pref_list_a.get_prefSC();
+  }
+
+  for (auto &it: non_proposing_partition) {
+    auto b = it.second;
+    auto pref_list_b = b->get_preference_list();
+    prefS[b] = pref_list_b.get_prefS();
+    prefSC[b] = pref_list_b.get_prefSC();
+  }
+
+  // while Q is not empty, do
+  // there is at least one vertex in the free list
+  while (not free_list.empty()) {
+    // arbitrary vertex in free list
+    auto u = remove_from_free_list(free_list, bookkeep_data);
+    const auto &u_pref_list = u->get_preference_list();
+    auto &u_data = bookkeep_data[u];
+    auto l = u_data.level;
+    bool star = u_data.star;
+
+    if (l < t) {
+      // if u^l has not exhausted PrefSC(u) then CriticalPropose(u^l, PrefSC(u),M,Q)
+      if (not u_data.is_exhausted()) {
+        critical_propose(free_list, u, prefSC[u], bookkeep_data, M);
+      }
+      else {
+        u_data.level += 1;
+        add_to_free_list(free_list, u);
```

Comment:
Why is the proposal index for u not reset here? Shouldn't it start proposal for the new level starting at the beginning?